packages/react-reconciler/src/ReactFiberScheduler.old.js

```diff
@@ -1836,9 +1842,20 @@ function scheduleWorkToRoot(fiber: Fiber, expirationTime): FiberRoot | null {
   return root;
 }
 
-export function warnIfNotCurrentlyBatchingInDev(fiber: Fiber): void {
+// in a test-like environment, we want to warn if dispatchAction() is
+// called outside of a TestUtils.act(...)/batchedUpdates/render call.
+// so we have a a step counter for when we descend/ascend from
+// act() calls, and test on it for when to warn
+// It's a tuple with a single value. Look for shared/createAct to
+// see how we change the value inside act() calls
+
+export function warnIfNotCurrentlyActingUpdatesInDev(fiber: Fiber): void {
   if (__DEV__) {
-    if (isRendering === false && isBatchingUpdates === false) {
+    if (
+      ReactShouldWarnActingUpdates.current === false &&
+      isRendering === false &&
+      isBatchingUpdates === false
```

Comment:
Let's check `isBatchingUpdates` first because that one will usually fail and short circuit the others. `ReactShouldWarnActingUpdates.current` check should come last because it's the most expensive.